### PR TITLE
feat: add key action to table in markdown

### PR DESCRIPTION
**Pull request by seonim-ryu**
### Please check if the PR fulfills these requirements
- [x] It's the right issue type on the title
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [ ] It does not introduce a breaking change or has a description of the breaking change

### Use case 1: tab key

Pressing the `tab` key in the markdown table moves the cursor cell by cell.

#### Before

* `tab` : The text is indented in the cell.
* `shift + tab` : The text is outdented in the cell.

tab-current

#### After

* `tab` :
  * If it is not the last cell in a row, pressing `tab` moves to the next cell.
  * For the last cell in a row, pressing `tab` moves to the first cell in the next row.
  * If you press `tab` in the last cell of the table, he cursor is positioned outside the table (ending point).
* `shift + tab` :
  * If it is not the first cell in a row, pressing `shift + tab` moves to the previous cell.
  * For the first cell in a row, pressing `shift + tab` moves to the last cell in the previous row.
  * If you press `shift + tab` in the first cell of the table, the cursor is positioned outside the table (starting point).

tab

### Use case 2 : enter key

Pressing the `enter key` in the markdown table adds the syntax for the new row.

#### Before

* Cell of table head : Table syntax is broken.
* Cells in delimiter row : Table syntax is broken.
* Cell of table body : A newline is added to the text in the cell and rendered with the remaining columns filled.

enter-current

#### After

* Cell of table head : Table syntax is broken.
* Cells in delimiter row : A new row with as many columns as the current number of columns is added.
* Cell of table body :
  * A new row with as many columns as the current number of columns is added.
  * When adding a new row in the last row, if the current row is empty, remove the current row.

enter

---
Thank you for your contribution to TOAST UI product. 🎉 😘 ✨

### CODE REVIEW

File: apps/editor/src/js/markdownEditor.js [hunk 1]
```diff
@@ -493,6 +494,104 @@ class MarkdownEditor extends CodeMirrorExt {
...
+  _addTableRowByCell(cell) {
+    const { cm } = this;
+    const line = getMdStartLine(cell);
+    const nextRow = this.toastMark.findNodeAtPosition([line + 1, 1]);
+
+    const currentLineText = cm.getLine(line - 1);
+    const rowStr = createTableRow(cell.parent);
+
+    if ((nextRow && nextRow.type === 'tableRow') || currentLineText !== rowStr) {
+      cm.setCursor(line - 1, getMdEndCh(cell.parent));
```
Comment (lja1018) [inline apps/editor/src/js/markdownEditor.js:529]: `const { parent } = cell` 는 어떤가요??

File: apps/editor/src/js/utils/markdown.js [hunk 2]
```diff
@@ -129,3 +141,17 @@ export function setChPos(originPos, newCh) {
...
+  let result = '|';
+
+  while (cell) {
+    if (isTableCellNode(cell)) {
+      result += '  |';
+    }
+    cell = cell.next;
+  }
+
+  return result;
```
Comment (js87zz) [inline apps/editor/src/js/utils/markdown.js:156]: 아래처럼 더 간결하게 변경할 수 있겠네요~
```js
const columnLen = row.parent.parent.columns.length;
let result = '|';

for (let i =0; i < columnLen; i += 1) {
  result += '  |';
}
return result;
```

File: apps/editor/src/js/markdownEditor.js [hunk 3]
```diff
@@ -493,6 +494,104 @@ class MarkdownEditor extends CodeMirrorExt {
...
+      this._addTableRowByCell(cellNode);
+    } else {
+      this.cm.execCommand('newlineAndIndentContinueMarkdownList');
+    }
+  }
+
+  _addTableRowByCell(cell) {
+    const { cm } = this;
+    const line = getMdStartLine(cell);
+    const nextRow = this.toastMark.findNodeAtPosition([line + 1, 1]);
```
Comment (js87zz) [inline apps/editor/src/js/markdownEditor.js:523]: `findNodeAtPosition`로 탐색하면 루트 노드부터 탐색하니까 `cell.parent.next`로 찾는게 더 좋지 않을까요?

File: apps/editor/src/js/markdownEditor.js [hunk 4]
```diff
@@ -493,6 +494,104 @@ class MarkdownEditor extends CodeMirrorExt {
...
+  }
+
+  _moveCursorNextCell(cell) {
+    let line = getMdStartLine(cell);
+    let ch = getMdEndCh(cell) + 2;
+
+    if (cell.next) {
+      ch = getMdEndCh(cell.next);
+    } else {
+      const nextRow = this.toastMark.findNodeAtPosition([line + 1, 1]);
```
Comment (js87zz) [inline apps/editor/src/js/markdownEditor.js:566]: 여기도 `cell.parent.next`로 변경하는게 좋을 것 같습니다~!

File: apps/editor/src/js/markdownEditor.js [hunk 5]
```diff
@@ -493,6 +494,104 @@ class MarkdownEditor extends CodeMirrorExt {
...
+
+      if (prevRow && isTableRowNode(prevRow)) {
+        line = line - 1;
+        ch = getMdEndCh(prevRow.lastChild);
+      }
+    }
+
+    this.cm.setCursor({ line: line - 1, ch: ch - 1 });
+  }
+
```
Comment (js87zz) [inline apps/editor/src/js/markdownEditor.js:594]: 키맵핑 부분은 `command`로 별도 분리하는게 더 깔끔하지 않을까요?

File: apps/editor/src/js/utils/markdown.js (hunk 2 above)
Comment (seonim-ryu) [reply apps/editor/src/js/utils/markdown.js:156]: `columns` 정보가 있었군요. 간단하게 수정할 수 있겠네요ㅎ

File: apps/editor/src/js/markdownEditor.js (hunk 3 above)
Comment (seonim-ryu) [reply apps/editor/src/js/markdownEditor.js]: 델리미터 행의 경우 `cell.parent.next` 값이 `null`이어서 다음과 같이 수정합니다.
```js
const nextRow = !parent.next && parent.parent.type === 'tableHead' ? parent.parent.next.firstChild : parent.next;
```

File: apps/editor/src/js/markdownEditor.js (hunk 4 above)
Comment (seonim-ryu) [reply apps/editor/src/js/markdownEditor.js]: 델리미터 행의 경우 `cell.parent.next` 값이 `null`이어서 다음과 같이 수정합니다.
```js
const nextRow = !parent.next && parent.parent.type === 'tableHead' ? parent.parent.next.firstChild : parent.next;
```

File: apps/editor/src/js/markdownEditor.js (hunk 5 above)
Comment (seonim-ryu) [reply apps/editor/src/js/markdownEditor.js]: 커맨드로 분리해서 적용한 커밋입니다ㅎㅎ

> https://github.com/nhn/tui.editor/pull/1126/commits/c15692cfb89b96ab1a7bdec9f705e74a59c27e47

**Review by js87zz — APPROVED** [review]

**Pull request merged**